File: Labtwo/PHYTHON/pr2.py
import re

def is_valid_username(username):
    """Проверка валидности имени пользователя"""
    # Проверка длины
    if len(username) < 6 or len(username) > 30:
        return False
    
    # Не может начинаться/заканчиваться точкой
    if username.startswith('.') or username.endswith('.'):
        return False
    
    # Не может содержать две точки подряд
    if '..' in username:
        return False
    
    # Проверка допустимых символов через регулярное выражение
    return bool(re.fullmatch(r'[a-zA-Z0-9.*]+', username))

def normalize_email(email):
    """Нормализация email"""
    # Поиск @
    at_pos = email.find('@')
    if at_pos == -1:
        return None  # Некорректный email
    
    # Разделение на части
    username = email[:at_pos]
    domain = email[at_pos:]
    
    normalized = []
    ignore_after_star = False
    
    # Обработка каждого символа
    for c in username:
        if c == '*':
            ignore_after_star = True
            continue
        
        # Добавляем символ если не точка и не после звездочки
        if not ignore_after_star and c != '.':
            normalized.append(c.lower())  # В нижний регистр
    
    # Возвращаем нормализованный email
    return ''.join(normalized) + domain.lower()

# Основная программа
emails_input = input("Enter emails (separated by commas): ")
# Разделение ввода по запятым и очистка от пробелов
emails = [email.strip() for email in emails_input.split(',') if email.strip()]

# Множество для хранения уникальных email
unique_emails = set()

# Обработка каждого email
for email in emails:
    at_pos = email.find('@')
    if at_pos != -1 and is_valid_username(email[:at_pos]):
        normalized = normalize_email(email)
        if normalized:
            unique_emails.add(normalized)  # Добавление в множество

# Вывод результата
print(f"Number of unique emails: {len(unique_emails)}")

File: Labtwo/TYPE_SKR/pr2.ts
// Проверка валидности имени пользователя
function isValidUsername(username: string): boolean {
    // Проверка длины
    if (username.length < 6 || username.length > 30) {
        return false;
    }

    // Не может начинаться/заканчиваться точкой
    if (username.startsWith('.') || username.endsWith('.')) {
        return false;
    }

    // Не может содержать две точки подряд
    if (username.includes('..')) {
        return false;
    }

    // Проверка допустимых символов через регулярное выражение
    return /^[a-zA-Z0-9.*]+$/.test(username);
}

// Нормализация email
function normalizeEmail(email: string): string | null {
    // Поиск @
    const atPos = email.indexOf('@');
    if (atPos === -1) {
        return null;  // Некорректный email
    }

    // Разделение на части
    const username = email.substring(0, atPos);
    const domain = email.substring(atPos);

    let normalized = '';
    let ignoreAfterStar = false;

    // Обработка каждого символа
    for (const c of username) {
        if (c === '*') {
            ignoreAfterStar = true;
            continue;
        }

        // Добавляем символ если не точка и не после звездочки
        if (!ignoreAfterStar && c !== '.') {
            normalized += c.toLowerCase();  // В нижний регистр
        }
    }

    // Возвращаем нормализованный email
    return normalized + domain.toLowerCase();
}

// Основная программа
const input = prompt("Enter emails (separated by commas):") || "";
// Разделение ввода по запятым и очистка от пробелов
const emails = input.split(',').map(e => e.trim()).filter(e => e);

// Множество для хранения уникальных email
const uniqueEmails = new Set<string>();

// Обработка каждого email
for (const email of emails) {
    const atPos = email.indexOf('@');
    if (atPos !== -1 && isValidUsername(email.substring(0, atPos))) {
        const normalized = normalizeEmail(email);
        if (normalized) {
            uniqueEmails.add(normalized);  // Добавление в множество
        }
    }
}

// Вывод результата
console.log(`Number of unique emails: ${uniqueEmails.size}`);
